assets/js/googlesitekit/datastore/site/html.js

```diff
@@ -0,0 +1,217 @@
+import invariant from 'invariant';
+
+/**
+ * WordPress dependencies
+ */
+import { isURL, addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { STORE_NAME } from './constants';
+import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
+
+const { createRegistryControl } = Data;
+
+const fetchHTMLForURLStore = createFetchStore( {
+	baseName: 'getHTMLForURL',
+	controlCallback: async ( { url } ) => {
+		const fetchHTMLOptions = {
+			credentials: 'omit',
+		};
+		const fetchHTMLQueryArgs = {
+			// Indicates a tag checking request. This lets Site Kit know not to output its own tags.
+			tagverify: 1,
+			// Add a timestamp for cache-busting.
+			timestamp: Date.now(),
+		};
+		const response = await fetch( addQueryArgs( url, fetchHTMLQueryArgs ), fetchHTMLOptions );
+		if ( ! response.ok ) {
+			throw {
+				code: response.statusText,
+				message: response.statusText,
+				data: { status: response.status },
+			};
+		}
+		const html = await response.text();
+
+		return html;
+	},
+	reducerCallback: ( state, htmlForURL, { url } ) => {
+		return {
+			...state,
+			htmlForURL: {
+				...state.htmlForURL,
+				[ url ]: htmlForURL,
+			},
+		};
+	},
+	argsToParams: ( url ) => {
+		invariant( isURL( url ), 'a valid url is required to fetch HTML.' );
+		return { url };
+	},
+} );
+
+// Actions
+const RESET_HTML_FOR_URL = 'RESET_HTML_FOR_URL';
+const WAIT_FOR_HTML_FOR_URL = 'WAIT_FOR_HTML_FOR_URL';
+
+export const BASE_INITIAL_STATE = {
+	htmlForURL: {},
+};
+
+const baseActions = {
+	/**
+	 * Resets the HTML for a given URL.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @param {string} url URL for which the HTML should be reset.
+	 * @return {Object} Redux-style action.
+	 */
+	*resetHTMLForURL( url ) {
+		const { dispatch } = yield Data.commonActions.getRegistry();
+
+		yield {
+			payload: { url },
+			type: RESET_HTML_FOR_URL,
+		};
+
+		return dispatch( STORE_NAME ).invalidateResolutionForStoreSelector( 'getHTMLForURL' );
+	},
+	/**
+	 * Waits for HTML for to be resolved for the given account URL.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @param {string} url URL for which to fetch HTML.
+	 * @yield {Object} Redux-style action.
+	 */
+	*waitForHTMLForURL( url ) {
+		yield {
+			payload: { url },
+			type: WAIT_FOR_HTML_FOR_URL,
+		};
+	},
+};
+
+const baseControls = {
+	[ WAIT_FOR_HTML_FOR_URL ]: createRegistryControl( ( registry ) => ( { payload: { url } } ) => {
+		// Select first to ensure resolution is always triggered.
+		registry.select( STORE_NAME ).getHTMLForURL( url );
+		const isHTMLForURLLoaded = () => ( registry.select( STORE_NAME ).hasFinishedResolution( 'getHTMLForURL', [ url ] ) );
+
+		if ( isHTMLForURLLoaded() ) {
+			return;
+		}
+
+		return new Promise( ( resolve ) => {
+			const unsubscribe = registry.subscribe( () => {
+				if ( isHTMLForURLLoaded() ) {
+					unsubscribe();
+					resolve();
+				}
+			} );
+		} );
+	} ),
```

Comment:
I just found out that it is possible to do this using an existing method:
```suggestion
		return registry.__experimentalResolveSelect( STORE_NAME ).getHTMLForURL( url );
```

This also has the added benefit of returning the resolved value from the selector when it completes.

E.g.

```js
const existingTag = yield actions.waitForExistingTag();
```

I'm working on an issue to cover a larger refactoring with this so let's keep it to here for now.